writeengine/dictionary/we_dctnry.cpp

```diff
@@ -745,6 +750,362 @@ int Dctnry::insertDctnry2(Signature& sig)
+
+      if (m_logger)
+      {
+        std::ostringstream oss;
+        oss << "Add dictionary extent OID-" << m_dctnryOID << "; DBRoot-" << m_dbRoot << "; part-"
+            << m_partition << "; seg-" << m_segment << "; hwm-" << m_curFbo << "; LBID-" << startLbid
+            << "; file-" << m_segFileName;
+        m_logger->logMsg(oss.str(), MSGLVL_INFO2);
+      }
+
+      m_curLbid = startLbid;
+
+      // now seek back to the curFbo, after adding an extent
+      // @bug5769 For uncompressed only;
+      // ChunkManager manages the file offset for the compression case
+      if (m_compressionType == 0)
+      {
+#ifdef PROFILE
+        Stats::startParseEvent(WE_STATS_PARSE_DCT_SEEK_EXTENT_BLK);
+#endif
+        long long byteOffset = m_curFbo;
+        byteOffset *= BYTE_PER_BLOCK;
+        RETURN_ON_ERROR(setFileOffset(m_dFile, byteOffset));
+#ifdef PROFILE
+        Stats::stopParseEvent(WE_STATS_PARSE_DCT_SEEK_EXTENT_BLK);
+#endif
+      }
+    }
+    else
+    {
+      // LBIDs are numbered collectively and consecutively within an
+      // extent, so within an extent we can derive the LBID by simply
+      // incrementing it rather than having to go back to BRM to look
+      // up the LBID for each FBO.
+      m_curLbid++;
+    }
+
+#ifdef PROFILE
+    Stats::startParseEvent(WE_STATS_PARSE_DCT);
+#endif
+    m_curBlock.lbid = m_curLbid;
+
+    //..."found" flag indicates whether the string was already found
+    //   "or" added to the end of the previous block.  If false, then
+    //   we need to add the string to the new block.
+    if (!found)
+    {
+      RETURN_ON_ERROR(insertDctnry2(curSig));  // m_freeSpace updated!
+      m_curBlock.state = BLK_WRITE;
+      memcpy(pOut + outOffset, &curSig.token, 8);
+      outOffset += 8;
+      startPos++;
+
+      //...Add string to cache, if we have not exceeded cache limit
+      if ((m_arraySize < MAX_STRING_CACHE_SIZE) && (curSig.size <= MAX_SIGNATURE_SIZE))
+      {
+        addToStringCache(curSig);
+      }
+    }
+  }  // if next
+
+  return NO_ERROR;
+}
+
+/*******************************************************************************
+ * Description:
+ * Used by bulk import to insert batch of parquet strings into this store file.
+ * Function assumes that the file is already positioned to the current block.
+ *
+ * PARAMETERS:
+ *    input
+ *       columnData - arrow array containing input strings
+ *       startRowIdx - start position for current batch parquet data
+ *       totalRow - number of rows in "buf"
+ *       col - column of strings to be parsed from "buf"
+ *    output
+ *       tokenBuf   - tokens assigned to inserted strings
+ *
+ * RETURN:
+ *    success     - successfully write the header to block
+ *    failure     - it did not write the header to block
+ ******************************************************************************/
+int Dctnry::insertDctnryParquet(std::shared_ptr<arrow::Array> columnData, int startRowIdx,
```

Comment:
This function's control flow is parametrized by: nullability of a column and fixed/variable size property. Unfortunately these params checks were moved into the loop and happens multiple times that shouldn't affect the perf significantly b/c of speculative execution but IMHO this should be re-written with either templates or classes with checks moved outside the loop.